events/events.go

```diff
@@ -17,112 +17,125 @@ import (
 const (
 	EventResponderCreate = "responder.create"
 	EventCommentCreate   = "comment.create"
+	EventPushQueueCreate = "push_queue.create"
 )
 
-func ListenForEvents() {
+const (
+	eventQueueBatchSize   = 1
+	eventMaxAttempts      = 3
+	waitDurationOnFailure = time.Second * 10
+	pgNotifyTimeout       = time.Minute
+)
+
+type Config struct {
+	UpstreamConf api.UpstreamConfig
+}
 
-	logger.Infof("Started listening for events")
+func ListenForEvents(ctx context.Context, config Config) {
+	logger.Infof("started listening for database notify events")
 
 	// Consume pending events
-	consumeEvents()
+	consumeEventsUntilEmpty(ctx, config)
 
 	pgNotify := make(chan bool)
-	go listenToPostgresNotify(pgNotify)
+	go listenToPostgresNotify(ctx, pgNotify)
 
 	for {
 		select {
 		case <-pgNotify:
-			consumeEvents()
+			consumeEventsUntilEmpty(ctx, config)
 
-		case <-time.After(60 * time.Second):
-			consumeEvents()
+		case <-time.After(pgNotifyTimeout):
+			consumeEventsUntilEmpty(ctx, config)
 		}
 	}
 }
 
-func listenToPostgresNotify(pgNotify chan bool) {
-	ctx := context.Background()
-
+func listenToPostgresNotify(ctx context.Context, pgNotify chan bool) {
 	conn, err := db.Pool.Acquire(ctx)
 	if err != nil {
-		logger.Errorf("Error creating database pool: %v", err)
+		logger.Errorf("error creating database pool: %v", err)
 	}
 
 	_, err = conn.Exec(ctx, "LISTEN event_queue_updates")
 	if err != nil {
-		logger.Errorf("Error listening to database notify: %v", err)
+		logger.Errorf("error listening to database notify: %v", err)
 	}
 
 	for {
 		_, err = conn.Conn().WaitForNotification(ctx)
 		if err != nil {
-			logger.Errorf("Error waiting for database notifications: %v", err)
+			logger.Errorf("error waiting for database notifications: %v", err)
 		}
 
 		pgNotify <- true
 	}
-
 }
 
-func consumeEvents() {
-
-	consumeEvent := func() error {
-		var event api.Event
-
-		tx := db.Gorm.Begin()
+func consumeEvents(ctx context.Context, config Config) error {
+	tx := db.Gorm.WithContext(ctx).Begin()
+	if tx.Error != nil {
+		return fmt.Errorf("error initiating db tx: %w", tx.Error)
+	}
 
-		selectEventsQuery := `
-			SELECT * FROM event_queue
-			WHERE
-				attempts <= 3 OR ((now() - last_attempt) > '1 hour'::interval)
-			FOR UPDATE SKIP LOCKED
-			LIMIT 1
-		`
-		err := tx.Raw(selectEventsQuery).First(&event).Error
-		if err != nil {
-			// Commit the transaction in case of no records found to prevent
-			// creating dangling connections and to release the locks
-			tx.Commit()
-			return err
-		}
+	selectEventsQuery := fmt.Sprintf(`
+		SELECT * FROM event_queue
+		WHERE attempts <= %d OR ((now() - last_attempt) > '1 hour'::interval)
+		FOR UPDATE SKIP LOCKED
+		LIMIT %d
+	`, eventMaxAttempts, eventQueueBatchSize)
```

Comment:
Use named args here instead of fmt.Sprint

Reply:
Although it's a dynamically generated query, it's really a static query in essence. The injected variables do not change over the execution of the program & are also in our control. 

I don't think it's necessary to be injected through gorm.

Reply:
We might have to expose the `attempts` and `limit` to the users later on, so have to sanitize the input before sending it to the db

Reply:
ah I see